### ISSUE TITLE
Problem in Dropdown in Bootstrap 4.2.1

### ISSUE THREAD

**Issue by afuno** [twbs/bootstrap#27903]
Bootstrap 4.2.1 / macOS 10.14.2 / Chrome 71.0.3578.98 / Safari 12.0.2

Dropdown works: https://codepen.io/anon/pen/oJZKXW
Dropdown doesn't work: https://codepen.io/anon/pen/BvWXNO

> Uncaught DOMException: Failed to execute 'querySelector' on 'Document': '/posts' is not a valid selector.

The problem arises due to the presence of the link.
In the application there are many places where the link in this place is “just like that”.

I think... it is necessary to fix it not on my side.
What do you think?

**Comment by sts-ryan-holton** [comment]
I'm unable to reproduce after upgrading from 4.1.3 to 4.2.1 & seems fine in the docs?

**Comment by PixemWeb** [comment]
I had the same issue. What I did to resolve it was looked over the code in the bootstrap.bundle.js from the 4.1.3 version, found the code that caused the issue in 4.2.1 and then replaced it.

Basically in the 4.1.3 version, I went to line 141 which doesn't cause an error.

```
    getSelectorFromElement: function getSelectorFromElement(element) {
        var selector = element.getAttribute('data-target');

        if (!selector || selector === '#') {
          selector = element.getAttribute('href') || '';
        }

        try {
          return document.querySelector(selector) ? selector : null;
        } catch (err) {
          return null;
        }
      },
```

I then copied that version over to the 4.2.1 version and replaced it at line number 141. The code below is what causes the issue and needs to be replaced with the snippet above.

```
    getSelectorFromElement: function getSelectorFromElement(element) {
      var selector = element.getAttribute('data-target');

      if (!selector || selector === '#') {
        var hrefAttr = element.getAttribute('href');
        selector = hrefAttr && hrefAttr !== '#' ? hrefAttr.trim() : '';
      }

      return selector && document.querySelector(selector) ? selector : null;
    },
```
Just make sure to comment out the current code that you're replacing just in case it causes issues and test accordingly. 

This was a quick fix and while I checked for other references of the code usage, there's about 12 in the file, it didn't seem to break anything. Note: I haven't tested with other bootstrap features like the Carousel. So make sure to test it with the various features you implement.

Hope this helps.

**Comment by hiro20v** [comment]
Probably the same problem is occurring.

[Maybe from this commitment](https://github.com/twbs/bootstrap/commit/a689120fd2af44bfb317f899b42037c46bfcd2fe#diff-9190923e07de66e65c3e197c2e8b01fe)

Occurrence code.
```
<html lang="en">

<head>
  <meta charset="utf-8">
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.2.1/css/bootstrap.min.css" integrity="sha384-GJzZqFGwb1QTTN6wy59ffF1BuGJpLSa9DkKMp0DgiMDm4iYMj70gZWKYbI706tWS" crossorigin="anonymous">
</head>

<body>

  <a href="/home" class="nav-link"></a>

  <div class="bd-example">
    <nav id="navbar-example2" class="navbar navbar-light bg-light">
      <a class="navbar-brand" href="#">Navbar</a>
      <ul class="nav nav-pills">
        <li class="nav-item">
          <a class="nav-link" href="#fat">@fat</a>
        </li>
      </ul>
    </nav>
    <div data-spy="scroll" data-target="#navbar-example2" data-offset="0" class="scrollspy-example">
      <h4 id="fat">@fat</h4>
      <p></p>
    </div>
  </div>

  <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
  <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.2.1/js/bootstrap.min.js" integrity="sha384-B0UglyR+jN6CkvvICOB2joaf5I4l3gm9GU6Hc1og6Ls7i6U/mkkaduKaBhlAXv9k" crossorigin="anonymous"></script>
  <script type="text/javascript">
    $('[data-spy="scroll"]').each(function() {
      var $spy = $(this).scrollspy('refresh')
    })
  </script>
</body>

</html>

```

**Comment by mattez** [comment]
My genius colleague hotfixed this by adding data attribute `data-target` to dropdown-toggle and coresponding ID to dropdown-menu. I don't know why, but it works.
See here:
https://codepen.io/mattez/pen/romzmO
(I forked and edit not working example from start of this thread)

Why is this starts working? There is no mention in the docs about any `data-target` in the Dropdown component (only in 'collapse')

**Comment by ankurk91** [comment]
@mattez
You solution works but have one caveat, if user clicks on the same anchor tag again, it does not close (toggle) the dropdown.

**Comment by ronit-mukherjee** [comment]
@mattez 
I am seeing a blocking error on the load of the bootstrap.
The error appears in the **bootstrap.min.js**
**Bootstrap Version:-** 4.2.1

**The error is:-** 
Uncaught DOMException: Failed to execute 'querySelector' on 'Document': '/c/10013' is not a valid selector.

![bootstrap-error](https://user-images.githubusercontent.com/39576708/50503084-34da5800-0a8a-11e9-8720-d30f0599052e.PNG)
![bootstrap-error2](https://user-images.githubusercontent.com/39576708/50503154-a9ad9200-0a8a-11e9-811b-e618732707cc.PNG)

**Comment by Johann-S** [comment]
A temporary fix would be to add a `data-target` attribute on your link to avoid that error, for example:

```html
<a href="/profile" data-target="#" class="nav-link">Link</a>

**Comment by mattez** [comment]
@Johann-S If I understand this well, will it be fix not earlier then in version [v4.3](https://github.com/twbs/bootstrap/projects/16)?

**Comment by Johann-S** [comment]
exactly @mattez but we plan to cut v4.3 in shorter time compared to v4.2

**Comment by DBenS** [comment]
I'm having the same problem using the following code in ASP.NET:

```			<li Class="nav-item dropdown">
              <a runat="server" Class="nav-link dropdown-toggle" href="#" id="AdmFuncs" role="button" data-target="#" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">Arquivos</a>
					<div class="dropdown-menu" aria-labelledby="navbarDropdown" >
						<a class="dropdown-item" href="/AdminActions/Categorias.aspx">Categorias</a>
						<a class="dropdown-item" href="/AdminActions/Fabricantes.aspx">Fabricantes</a>
						<a class="dropdown-item" href="/AdminActions/Personagem.aspx">Personagens</a>
						<a class="dropdown-item" href="/AdminActions/EstadoUso.aspx">Estados</a>
						<div class="dropdown-divider"></div>
						<a class="dropdown-item" href="/AdminActions/Recepcao.aspx">E-mail Receptor</a>
			        </div>
            </li>

```

When I jump to any link in a subfolder, the problem appears...

DATA-TARGET did not solve it.

Best regards

**Comment by Homelux** [comment]
Big porblem! in bootstrap 4.2.1 does NOT work in navbar dropdown menu.

**Comment by florianlacreuse** [comment]
Just to let you know that since `href` with relative link is useless in most cases with dropdown you could just use `href="#"` to fix it too, just like in the [navbar documentation](https://getbootstrap.com/docs/4.2/components/navbar/#supported-content).

**Comment by Homelux** [comment]
> Just to let you know that since `href` with relative link is useless in most cases with dropdown you could just use `href="#"` to fix it too, just like in the [navbar documentation](https://getbootstrap.com/docs/4.2/components/navbar/#supported-content).

href="#" NOT fix, in Joomla very often you need to make a parent menu item for a category, and you just need a link! 

not href="#"

**Comment by florianlacreuse** [comment]
In this case use fix provided by @Johann-S, no big deal over here.

**Comment by Johann-S** [comment]
@Homelux you should add a `data-target` attribute with the following value: `#` and you'll be able to keep your link.

There is an example here: https://github.com/twbs/bootstrap/issues/27903#issuecomment-450292043

**Comment by Homelux** [comment]
data-target for parent link or child link ?

**Comment by Johann-S** [comment]
It depends which one trigger the dropdown

**Comment by Homelux** [comment]
parent link - open dropdown

**Comment by Johann-S** [comment]
so you should add `data-target` to your parent link

**Comment by Homelux** [comment]
thanks

and please make hover dropdown for navbar as foundation, very need!

**Comment by Homelux** [comment]
data-target dont work

http://stihidom.ru/

**Comment by Johann-S** [comment]
Change your link by:

```html
<a class="nav-link dropdown-toggle" href="#" data-toggle="dropdown" data-target="#" aria-expanded="false">Наши услуги</a>
```
it'll work

**Comment by Homelux** [comment]
not, i need /uslugi for seo, not href="#"

parent link -> article category with link /uslugi
child links -> articles from this category with links 

/uslugi/article1
/uslugi/article2
/uslugi/article3

therefore it is important parent link /uslugi for seo, not href="#"

**Comment by emage** [comment]
@Johann-S  The fix does not work for me as well.

My link:

`<a class="nav-link dropdown-toggle" href="/company/contactus.aspx" data-target="#" role="button" id="contact-phone-listing" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">Contact Us</a>`

Our code requires a link in the HREF.

**Comment by Homelux** [comment]
Yes, our code requires a link in the HREF.

It worked great in version 4.1.3.

**Comment by Johann-S** [comment]
@Homelux if you want that, you have to specified a `id` attribute on the dropdown menu and fill the `data-target` attribute with the selector for your menu.

See my example:

![image](https://user-images.githubusercontent.com/1689750/50886030-aa7de680-13ef-11e9-83a1-3a18b846120d.png)

the dropdown menu has the id `myDropdown` and the link's `data-target` is equal to `#myDropdown` and it works.

Same use case for @emage

**Comment by Saintless** [comment]
@Johann-S THANK YOU!!! That fixed it for me.

I had my menu set up like this, which was causing the error from above:
```
<li class="nav-item dropdown">
	<a class="nav-link dropdown-toggle" href="/Summary/Commercial" data-target="#" data-toggle="dropdown" id="navToggleCommercial" role="button" aria-haspopup="true" aria-expanded="false">
		Commercial
	</a>
	<div class="dropdown-menu" aria-labelledby="navToggleCommercial">
		<a class="dropdown-item" href="/Summary/Dashboard">Dashboard</a>
	</div>
</li>
```

Essentially, I had to move the ID to the dropdown div, and reference it with data-target.
```
<li class="nav-item dropdown">
	<a class="nav-link dropdown-toggle" href="/Summary/Commercial" data-target="#navToggleCommercial" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">
		Individual
	</a>
	<div class="dropdown-menu" id="navToggleCommercial">
		<a class="dropdown-item" href="/Summary/Dashboard">Dashboard</a>
	</div>
</li>
```

**Comment by bjarnef** [comment]
I have the same issue and I don't get the console error, when adding an unique id for the dropdown e.g. `myDropdown-1234`. However this doesn't seem to fix the issue to toggle the nav item - it opens to dropdown, but it doesn't close on next click.

When I just set `href="#" it works, but it seems only to work with two levels (first level + dropdown).

**Comment by bjarnef** [comment]
@Johann-S it is the same issue, but the "fix" with adding the id only open the dropdown, but doesn't close it on next click.

![image](https://user-images.githubusercontent.com/2919859/51036592-e2328d00-15ad-11e9-9e41-bcc15c0b0e2c.png)

Using href="#" works and open only child dropdown. When using the suggested solution with the id, it open also descendant menu levels.

**Comment by Johann-S** [comment]
@bjarnef not totally because you're using multi dropdown level which isn't supported by Bootstrap

**Comment by bjarnef** [comment]
@Johann-S yes, but it doesn't work with a single level either, when using the approach with an id, only when changing to `href="#"`

Do you have an example (on CodePen or JSFiddle) with Bootstrap v4.2.1, where it close the dropdown after second click?

**Comment by DBenS** [comment]
@Johann-S , no one of the above solutions function with ASP.NET using Masterpage - and the 4.1.3 version was functioning very well.

See the code below. It functions at the main page (the first page the website runs) but when I entry in a DropDown NAV-Item, the "selection function" really stops.

```
<li id="Arquivos" Class="nav-item dropdown">
     <a runat="server" Class="nav-link dropdown-toggle" href="#" data-target="#Arquivos" id="AdmFuncs" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" >Arquivos</a>
	<div class="dropdown-menu" aria-labelledby="AdmFuncs" id="MyArqs" >
		<a class="dropdown-item" href="/AdminActions/Categorias.aspx">Categorias</a>
		<a class="dropdown-item" href="/AdminActions/SubCategoria.aspx">Sub-Categorias</a>
		<a class="dropdown-item" href="/AdminActions/EstadoUso.aspx">Estados</a>
		<div class="dropdown-divider"></div>
		<a class="dropdown-item" href="/AdminActions/Recepcao.aspx">E-mail Receptor</a>
	  </div>
</li>

```

I tried all the above solutions and all of them I had the same behavior.

Thank you for your attention.
Kindest regards,
David

**Comment by Johann-S** [comment]
@DBenS I cannot help you more than that, if it works on your first page maybe there is something wrong on your other pages.

Do not hesitate to ask around in the official Slack team or Stack Overflow. [See our readme for more details](https://github.com/twbs/bootstrap#community).

@bjarnef there is an issue open here : https://github.com/twbs/bootstrap/issues/27659 about multi level dropdown

**Comment by emage** [comment]
@Johann-S Thank you!  
Your updated example solved my issues.

Thanks!

**Comment by sheilasward** [comment]
The problem is that Dropdowns do not work in bootstrap 4.2 unless you have the javascript files at the bottom:
<script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.6/umd/popper.min.js" integrity="sha384-wHAiFfRlMFy6i5SRaxvfOCifBUQy1xHdJ/yoi7FRNXMRBu5WHdZYu1hA6ZOblgut" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.2.1/js/bootstrap.min.js" integrity="sha384-B0UglyR+jN6CkvvICOB2joaf5I4l3gm9GU6Hc1og6Ls7i6U/mkkaduKaBhlAXv9k" crossorigin="anonymous"></script>

**Comment by yourchoice** [comment]
Same for me

DOMException: Failed to execute 'querySelector' on 'Document': '/recettes' is not a valid selector.

O use in href="/recettes" for seo ...

**Comment by shresthabiswas** [comment]
> ```
> data-target="#"
> ```
Still not working!
![data-target](https://user-images.githubusercontent.com/11130716/52948275-211fe380-33a1-11e9-83a0-31e174e2d36e.JPG)

**Comment by denniseh** [comment]
Found that making id's start with a non-numeric character solved the "document.querySelector failed to execute" issue I was getting.

**Comment by nbinobied** [comment]
I'm having similar issue with 4.2.1 when adding Html.ActionLink, it causes the CSS/JS to break from the page. I'd appreciate a way to fix this.

I'm getting the following error:

> Uncaught DOMException: Failed to execute 'querySelector' on 'Document': '/ControlPanel' is not a valid selector.


```
                <li class="nav-item">
                    <a class="nav-link js-scroll-trigger" href="#services">Services</a>
                </li>
                <li class="nav-item">
                    <a class="nav-link js-scroll-trigger" href="#portfolio">Portfolio</a>
                </li>
                <li class="nav-item">
                    <a class="nav-link js-scroll-trigger" href="#about">About</a>
                </li>
                <li class="nav-item">
                    <a class="nav-link js-scroll-trigger" href="#team">Team</a>
                </li>
                <li class="nav-item">
                    <a class="nav-link js-scroll-trigger" href="#contact">Contact</a>
                </li>
                <li class="nav-item">
                    @Html.ActionLink("Control Panel", "Index", "ControlPanel", null, new { @class = "nav-link js-scroll-trigger" })
                </li>
```

**Comment by Johann-S** [comment]
Updating your version of Bootstrap to the latest release will fix that (v4.3.1)
